Cleaned up image classification cpp example

lebeg opened the pull request:
## Description ##

Cleaned up image classification cpp example.

## Checklist ##
### Essentials ###
- [x] Passed code style checking (`make lint`)
- [x] Changes are complete

### Changes ###
- Introduced mxnet build option
- Enabled build on mac
- Cleaned up cmake file
- Applied minor code improvements
- Applied code formatting

#### Code improvements ####
- std::string -> const std::string& as method or function parameters
- Handling buffer memory with std::unique_ptr<char[]> instead of raw pointer
- Use std::size_t instead of int for sizes where possible
- Fixed log phrasing and whitespace formatting
- Used ostream for logs instead of printf
- Used standard exit codes from main
- Added license header

### Tested ###
- Mac OS
- Ubuntu
- Windows

larroy commented on example/image-classification/predict-cpp/image-classification-predict.cc:
```diff
@@ -32,252 +48,254 @@ const mx_float DEFAULT_MEAN = 117.0;
+  }
+
+  ~BufferFile() {
+    if (buffer_) {
```
the check is not needed

larroy replied:
Better a smart pointer in any case?

lebeg replied:
Done

larroy commented on example/image-classification/predict-cpp/image-classification-predict.cc:
```diff
@@ -32,252 +48,254 @@ const mx_float DEFAULT_MEAN = 117.0;
 // Code from : https://github.com/pertusa/mxnet_predict_cc/blob/master/mxnet_predict.cc
-std::vector<std::string> LoadSynset(std::string synset_file) {
-    std::ifstream fi(synset_file.c_str());
+std::vector <std::string> LoadSynset(std::string synset_file) {
```
why space before template arg?

larroy commented on example/image-classification/predict-cpp/image-classification-predict.cc:
```diff
@@ -32,252 +48,254 @@ const mx_float DEFAULT_MEAN = 117.0;
+  }
 
-    std::vector<std::string> output;
+  std::vector <std::string> output;
```
is this your IDE settings?

lebeg replied:
Thanks, this was a weird default setting on a new machine

piiswrong commented:
Those keys are release signing keys. If you are not going to drive a release you don't need it

marcoabreu commented on CMakeLists.txt:
```diff
@@ -671,8 +672,7 @@ if(USE_CPP_PACKAGE)
   add_subdirectory(cpp-package)
 endif()
 
-# Problems on Mac OS X: 1. librt not available 2. mxnet built as MODULE library, which can't be linked.
```
Has this been resolved?

lebeg replied:
Yes

marcoabreu commented on KEYS:
```diff
@@ -480,3 +480,40 @@ IxR3jVTKye+UerEtN8yATW8CRIKO3IobUfLMDdPCLO7uzoW95cI35Y0l8JgK2NeU
 =E4W5
 -----END PGP PUBLIC KEY BLOCK-----
 
+pub   rsa2048 2018-02-07 [SC] [expires: 2020-02-07]
```
Since you're not going to make a release, I don't think that your key should be in here.

marcoabreu commented:
Could you please add on which platforms you have tested your modifications? Since we're not testing examples on CI, we'll have to test them ourselves.

lebeg commented:
Listed tested platforms in the description

marcoabreu commented:
Could you please also test Windows?

lebeg commented:
Added windows to tested platforms

marcoabreu commented on example/image-classification/predict-cpp/CMakeLists.txt:
```diff
@@ -1,34 +1,28 @@
+# ---[ OpenCV
+if(NOT USE_OPENCV OR NOT OpenCV_FOUND)
+  message(WARNING "\
+OpenCV should be enabled and found to build image classification example, skipping...")
```
This is a critical error, right?

lebeg replied:
The example will just not be build and it will not break the whole project configuration.

cjolivier01 replied:
This certainly is chopping out a lot of stuff.  WHat's the net change in static linking logic?

lebeg replied:
It is not about static, it's an early return from this cmake file. The example can not be built if there is no OpenCV and will not be added to the main project. That's it.

cjolivier01 replied:
isn’t the unittest static link flag removed?

cjolivier01 replied:
UNITTEST_STATIC_LINK is descriptive and editable.  Changing the check to MSVC at the actual interpretation point makes it difficult to see what's being done.
Also, I think MSVC has to build static because of dllimport issues

lebeg replied:
I can leave the variable itself but would rename it, because it's an example, not a unit test.

lebeg commented:
@marcoabreu @piiswrong @cjolivier01 @szha do you want to take another look and merge if possible?

marcoabreu commented:
I can't really help here, sorry.

cjolivier01 commented on CMakeLists.txt:
```diff
@@ -32,6 +32,7 @@ mxnet_option(USE_GPROF            "Compile with gprof (profiling) flag" OFF)
 mxnet_option(USE_CXX14_IF_AVAILABLE "Build with C++14 if the compiler supports it" OFF)
 mxnet_option(USE_VTUNE            "Enable use of Intel Amplifier XE (VTune)" OFF) # one could set VTUNE_ROOT for search path
 mxnet_option(ENABLE_CUDA_RTC      "Build with CUDA runtime compilation support" ON)
+mxnet_option(BUILD_CPP_EXAMPLES   "Build cpp examples" OFF)
```
default to on, please. Otherwise, people may check in/merge broken code.

lebeg replied:
Ok

cjolivier01 commented on example/image-classification/predict-cpp/image-classification-predict.cc:
```diff
@@ -1,283 +1,298 @@
+ */
+
 /*!
- *  Copyright (c) 2015 by Xiao Liu, pertusa, caprice-j
```
I don't think you can do this, can you? I was under the impression that license text could never be removed.
  @smarthi , this this ok?

lebeg replied:
Technically I'm not removing contributors, just naming them as a group. I followed the pattern in the rest of the project, but will be happy to change it back.

cjolivier01 replied:
is the guy’s name still there?

lebeg replied:
Yes, brought it back

cjolivier01 commented on example/image-classification/predict-cpp/image-classification-predict.cc:
```diff
@@ -1,283 +1,298 @@
 
 const mx_float DEFAULT_MEAN = 117.0;
 
+std::string trim(const std::string& input) {
```
make static, please

lebeg replied:
Ok

cjolivier01 commented:
Can you please summarize the "cleaning up" items in the description? It's a little hard to follow what the added value is from the diff, since it's a bit scattered.

lebeg commented:
@cjolivier01 The changes are already listed:

### Changes ###
* Introduced mxnet build option
* Enabled build on mac
* Cleaned up cmake file
* Applied minor code improvements
* Applied code formatting

lebeg commented:
@cjolivier01 please take another look, thanks!

cjolivier01 commented:
Can you please summarize the "cleaning up" items in the description? It's a little hard to follow what the added value is from the diff, since it's a bit scattered.

cjolivier01 commented:
Also, is this used in any of the unit tests or nightly tests?  If not, probably we should be running at least one test with it.

lebeg commented:
@cjolivier01 for now this is not tested on ci, but I will ingerate this as the base for valgrind memcheck tests

lebeg commented:
Added 'Code improvements' section for detailisation
